### ISSUE TITLE
spring_layout (aka fruchterman_reingold) documentation needs improvement

### ISSUE THREAD

**Issue by joseortiz3** [networkx/networkx#3489]
Background: In the spring_layout function, an attractive "spring force" proportional to the adjacency matrix (and proportional to the distance between nodes) exists between each pair of nodes. This is counteracted by a repulsive "anti-gravity" force inversely proportional to the distance-squared between nodes. 

[code](https://github.com/networkx/networkx/blob/d27930ce099d6f0edca8d5d91d5f3a8d9dc90c9d/networkx/drawing/layout.py#L528)
```python
displacement = np.einsum('ijk,ij->ik',
                                 delta,
                                 (k * k / distance**2 - A * distance / k))
```

Doing a simple estimation, this means the equilibrium distance ("Optimal distance") where these attractive and repulsive forces cancel-out currently is inversely proportional to the cube-root of the values of the adjacency matrix, which could be anything from zero to infinity. In simpler terms, multiplying the adjacency matrix by 8 results in half the equilibrium distance.

In contrast, the documentation suggests the distance is completely controlled by the parameter `k`, the inverse-spring-constant.

> k : float (default=None)
>         Optimal distance between nodes.  If None the distance is set to
>         1/sqrt(n) where n is the number of nodes.  Increase this value
>         to move nodes farther apart.

Thus I propose to normalize the adjacency matrix values for the purpose of this layout algorithm, so that k really is the equilibrium distance: this means simply adding something like  `A = A / np.mean(A[A!=0])` before calculations. (np.median doesn't work as nicely as np.mean for some reason, other schemes possible)

Thoughts?

**Comment by dschult** [comment]
I would much prefer to change the documentation to match the code than to change the code to match the docs.  This is because it reduces backward compatibility problems, thus perhaps saving many people time when they upgrade to the next version. What do you think?

**Comment by joseortiz3** [comment]
Thinking about what backwards compatibility problems this change could cause, I made a script:

```python
import networkx as nx
import matplotlib.pyplot as plt
import numpy as np

# Generate graph
A = np.array([[0, 1, 2, 0],
              [1, 0, 3, 0],
              [2, 3, 0, 4],
              [0, 0, 4, 0]])
A[A!=0] = 2**(3*(A[A!=0]-1))
G = nx.Graph(A)

# Spring layout positions
pos = nx.spring_layout(G, iterations=500, seed=0, fixed=[1], pos={1:(0,0)})

# Unit box
plt.hlines([-1,0,1], -1, 1, linestyle=':')
plt.vlines([-1,0,1], -1, 1, linestyle=':')

# Draw things
nx.draw_networkx_nodes(G, pos=pos, node_size=100)
nx.draw_networkx_edges(G, pos=pos)
nx.draw_networkx_edge_labels(G, pos=pos, edge_labels={(v[0],v[1]):v[2] for v in list(G.edges.data('weight'))})

# Tick marks
plt.title('No Normalization of A and one fixed node, no scale reference')
plt.xticks([-1,0,1],[-1,0,1])
plt.yticks([-1,0,1],[-1,0,1])
ax = plt.gca()
ax.tick_params(
    axis='both',
    which='both',
    bottom=True,
    left=True,
    labelbottom=True,
    labelleft=True)

plt.show()
```

I put the normalization in the code for `def _frutchterman_reingold`:

```python
    try:
        nnodes, _ = A.shape
    except AttributeError:
        msg = "fruchterman_reingold() takes an adjacency matrix as input"
        raise nx.NetworkXError(msg)

    # Normalize A so only k controls equilibrium distance
    A = A / np.mean(A[A!=0])

    if pos is None:
        # random initial positions
        pos = np.asarray(seed.rand(nnodes, dim), dtype=A.dtype)
```

I also made sure it works for `def _sparse_frutchterman_reingold`:

```python
try:
        from scipy.sparse import spdiags, coo_matrix
    except ImportError:
        msg = "_sparse_fruchterman_reingold() scipy numpy: http://scipy.org/ "
        raise ImportError(msg)

    # Normalize A so only k controls equilibrium distance
    A = A / np.median(A[A!=0])

    # make sure we have a LIst of Lists representation
    try:
        A = A.tolil()
```

For the following plots: 

- "with normalization" means with that code inserted, "No" means commented out.
- "with fixed point" means giving that `fixed` and `pos` arguments to `sparse_layout`, anything else means giving neither `fixed` nor `pos` arguments.
- if the "unit box" isn't shown (a dashed box from -1 to 1), I commented out the `hlines` and `vlines`

Without normalization and no fixed points and unit box, the current behavior is:
![No Normalization](https://user-images.githubusercontent.com/18585219/60135726-3d6ccc00-9757-11e9-9da7-741f2832fb03.png)

With normalization and no fixed points and unit box, the new behavior would be:
![With normalization](https://user-images.githubusercontent.com/18585219/60135766-4c537e80-9757-11e9-92d9-ea7f6955e236.png)

Without normalization and one fixed point and unit box, the current behavior disables rescaling:
![No normalization one fixed](https://user-images.githubusercontent.com/18585219/60135797-5bd2c780-9757-11e9-9393-bfc57da3f55e.png)

With normalization and one fixed point and unit box, the new behavior retains disabled rescaling but the normalization has its own kind of rescaling:
![With normalization one fixed](https://user-images.githubusercontent.com/18585219/60135841-7efd7700-9757-11e9-8cf8-9dcf722d2afb.png)

Currently, users do not usually see a result like above, because the plot itself is rescaled (if no unit box is drawn):
![No normalization and no scale](https://user-images.githubusercontent.com/18585219/60135881-989ebe80-9757-11e9-80d2-4ce200497454.png)
same with x ticks
![No normalization and no unit box](https://user-images.githubusercontent.com/18585219/60135927-bbc96e00-9757-11e9-9cd2-053b06e3b1fb.png)

**Comment by dschult** [comment]
Thanks for that summary. Let's see if I understand correctly:

Current unexpected behavior is:

- With no fixed nodes:
    - anti-gravity is not scaled nor impacted by changing A
    - threshold cutoffs are hard coded and thus not impacted by changing A
- With any number of fixed nodes:
    - scaling is disabled so scale of results differ greatly from no fixed node case.
    - the unexpected behaviors for no fixed nodes still apply to this case as well.

In addition it seems to me that the docs aren't clear about much of this. It is not clear that scaling edge weights (A) will affect spring force but not anti-gravity. And it is not clear that the hard coded thresholds even exist nor what their impact is. 

The reason we disable rescaling when fixed nodes are present is because rescaling would affect the positions of the fixed nodes. 

Questions:

- Is there a better way to handle scaling with fixed nodes?
- Is there a better way to handle the trade-off between spring force, anti-gravity and thresholds?

If we were designing the API from scratch what would be best? Would we add keywords for anti-gravity and thresholds? Would we allow turning on scaling with fixed nodes? This is not the first time we've adjusted the scaling part of the API so I want to try to get it right.  

BTW, I like the center_spring solution to the unconnected graphs issue. Your Petri Dish example is quite compelling and could be included as a drawing example in our gallery. :)

**Comment by joseortiz3** [comment]
You understand everything correctly! And thank you!

Let me type this out as I'm thinking about it. There are two scales to be aware of: position scale (multiplying the numerical values of `pos`) and plot scale (controlled entirely by matplotlib). So we only need to worry about `pos` scale.

Starting from the beginning, what are the ideal parameter descriptions for `_sparse_frutchterman_reingold` and `_frutcherman_reingold` (which generation `pos`)

- `A` : matrix, adjacency matrix : Measured in "adjacency" units. The length (measured in `pos`-units) of edge[i,j] will be [_proportional to_ / equal and independent of / inversely proportional to] the value of the `A`[i,j]. (I think proportional to is the most logical. The edge length is then `k` times `A`[i,j])

---> How to do it: Currently anti gravity = k^2/r^2 (r = distance), and spring force = A[i,j]*r / k. Thus equilibrium-distance[i,j] = k / (A[i,j])^(1/3). Instead, make spring force = r / (k*A[i,j]^3), equivalent to the transformation A -> 1 / (A^3).

- `k` : float, inverse spring constant : Scale factor for distance between edges. The equilibrium length (measured in `pos` units) of edge[i,j] will be equal to `k` times `A`[i,j]. This is the edge length for which anti-gravity repulsion and spring-force attraction are equal.

----> No work necessary

- `pos` : dictionary : initial positions to start the iterations (unchanged)

----> No work necessary

- `fixed` : collection : a collection of nodes to fix the positions of (unchanged) at the initial position

----> No work necessary

- `center_spring_k` : Attracts every node towards the origin with a spring force, causing all nodes to occupy a square area of length `center_spring_k` approximately, at which point anti-gravity and the center spring force are in equilibrium.

----> How to do it: Make this origin-attracting spring force have magnitude r / k_center, where k_center = radius of circle/square enclosing all nodes = N^(1/dim) * mean(A[A!=0]) * k (basic approximation)

P.S. That code snippet
        # enforce minimum distance of 0.01
        np.clip(distance, 0.01, None, out=distance)

I figured out this is just a ceiling on the force of anti-gravity, expressed in terms of distance. So this should be set to the minimum distance we would expect in the final output, not to an arbitrary number. The minimum distance would thus be on order of `k` * max(`A`), perhaps 0.1*k*max(`A`).

**Comment by dschult** [comment]
We are definitely computing position scale, not plot scale.

We are in danger of mis-implementing the algorithm. This is ```fruchterman_reingold``` after all. So, I think we should NOT choose A -> 1/A^3. Also, the length should not be proportional to entries in A. With fruchterman_reingold the A values give spring constants not distances.

Questions we need to address:

- should we scale the positions if there are any fixed nodes (which would change the position of those fixed nodes)?  I think the answer is no. If people want to fix a node at a position then we won't scale.
- Do we need other keyword parameters to allow adjustment of scales? (now we have k, A and scale. These should provide enough flexibility to determine the spring scale, anti-gravity scale and threshold cutoff scale...  maybe?)

**Comment by joseortiz3** [comment]
> "Do we need other keyword parameters to allow adjustment of scales? (now we have k, A and scale..."
`k`, `A`, and `scale` (for the wrapper function) determine two scales.

The total range of `pos` (max - min) is determined by 
- If there are no `fixed` positions, only `scale` (default behavior) 
- If there are any `fixed` positions: `k`, and the largest values of `A`.
- This behavior is not consistent.
- In my opinion, `scale` should default to `None` (no scaling), leaving matplotlib to do the plot scaling for default behavior (instead of matplotlib's scaling being a redundant second scaling).
- having any `fixed` `pos`itions can only be interpreted in one of two ways: Fix the node positions inside the iterations to either 1) the raw `pos` specified in `spring_layout` (wrapper) and then disable scaling even if the user requests it or 2) reverse-scale and reverse-center the user-specified `pos` (if `scale`=None and center=`origin` this does nothing), in which case the fixed positions _after_ scaling and centering will match the user's specified positions. This has the added bonus of guaranteeing that feeding any `pos` output by `spring_layout` back into `spring_layout` will remain unchanged, i.e. will remain equilibrium `pos`. 2) is clearly better.

The relative lengths of the edges are determined by:
- The inverse cube root of A
- This behavior is not intuitive. Although its nice to interpret `A` as spring constants, spring constants aren't intuitive nor useful. Edge lengths are useful and intuitive. History should not impede innovation.

**Comment by dschult** [comment]
Maybe you should write an article about the Ortiz Layout Algorithm.
Then we would definitely include these ideas. :)
